File: Lua/_Examples/01_basics.lua
-- ☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆
-- ☆ Author: ☆ MelodyHSong ☆
-- ☆ Language: Lua
-- ☆ File Name: 01_basics.lua
-- ☆ Date: 2026-08-25
-- ☆
-- ☆ Description: Demonstrates Lua syntax basics including variables, control flow,
-- ☆              string manipulation, functions, and 1-based array indexing.
-- ☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆

print("==========================================")
print("  ☆ Welcome to Lua Basics ☆")
print("==========================================")

-- 1. Variables & Types
local greeting = "Hello, Lua!"
local version = _VERSION
local isAwesome = true
local price = 19.99

print(string.format("Status: %s (Running on %s)", greeting, version))
print(string.format("Types: isAwesome=%s (%s), price=%.2f (%s)", 
    tostring(isAwesome), type(isAwesome), price, type(price)))

-- 2. Functions & Control Flow
local function evaluateScore(score)
    if score >= 90 then
        return "Rank: S - Outstanding!"
    elseif score >= 75 then
        return "Rank: A - Great job!"
    elseif score >= 50 then
        return "Rank: B - Passing."
    else
        return "Rank: F - Try again!"
    end
end

print("\n--- Control Flow Evaluation ---")
local testScores = { 95, 82, 60, 45 }
for _, score in ipairs(testScores) do
    print(string.format("Score: %d -> %s", score, evaluateScore(score)))
end

-- 3. 1-Based Indexing Demonstration
print("\n--- 1-Based Array Indexing ---")
local inventory = { "Health Potion", "Mana Elixir", "Iron Sword", "Magic Ring" }

print("First item (index 1): " .. inventory[1])
print("Total items (#inventory): " .. #inventory)

print("Inventory List:")
for index, item in ipairs(inventory) do
    print(string.format("  [%d] %s", index, item))
end

print("\nBasics example executed successfully!")

File: Elixir/_Examples/01_basics.exs
# ☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆
# ☆ Author: ☆ MelodyHSong ☆
# ☆ Language: Elixir
# ☆ File Name: 01_basics.exs
# ☆ Date: 2026-08-25
# ☆
# ☆ Description: Demonstrates Elixir basics including pattern matching,
# ☆              immutable data structures, atoms, lists, and maps.
# ☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆☆

IO.puts("==========================================")
IO.puts("  ☆ Welcome to Elixir Basics ☆")
IO.puts("==========================================")

# 1. Atoms, Variables & Immutability
status = :ok
message = "Hello, Elixir!"
version = System.version()

IO.puts("Status Atom: #{inspect(status)}")
IO.puts("Message: #{message} (Elixir v#{version})")

# 2. Pattern Matching with Tuples & Destructuring
{:ok, result_val} = {:ok, 42}
IO.puts("Pattern Matched Tuple Value: #{result_val}")

# Destructuring lists into Head and Tail
[head | tail] = [10, 20, 30, 40]
IO.puts("List Head: #{head}")
IO.puts("List Tail: #{inspect(tail)}")

# 3. Maps & Map Access
player = %{
  username: "MelodyHSong",
  level: 50,
  class: :archmage,
  active?: true
}

IO.puts("\n--- Map Data Access ---")
IO.puts("Username: #{player.username}")
IO.puts("Class: #{inspect(player.class)}")
IO.puts("Level: #{player.level}")

# Updating a map (creates a new map immutably)
updated_player = %{player | level: 51}
IO.puts("Updated Level (Immutable): #{updated_player.level}")
IO.puts("Original Level Unchanged: #{player.level}")

# 4. Anonymous Functions
double = fn x -> x * 2 end
add = &(&1 + &2) # Capture operator shorthand syntax

IO.puts("\n--- Anonymous Functions ---")
IO.puts("double(21) => #{double.(21)}")
IO.puts("add.(15, 25) => #{add.(15, 25)}")

IO.puts("\nBasics example executed successfully!")
